### PR TITLE
[kiosk] Yet another take on extensions

### PR DESCRIPTION
**Pull request by damirka**
## Description 

Another attempt to implement a fail-safe Extensions API with a minimal set of introduced changes.

## Test Plan 

Tests to be added along with a set of extensions.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

TBD

### CODE REVIEW

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (sblackshear) [inline crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:552]: I think it would be easier to read if we expose a `has_place` or `assert_place_permission` function (and similar for the other permissions) rather than doing the bit arithmetic here

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (sblackshear) [inline crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:554]: Since we check this in a few places, might be useful to have a helper function

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (sblackshear) [inline crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:508]:
```suggestion
    public fun set_allowed_types<Ext: drop>(
```

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (sblackshear) [inline crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:497]:
```suggestion
    public fun enable_extension<Ext: drop>(
```

+ noting that extensions are automatically enabled after `add_extension` will be clearer, I think.

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (sblackshear) [inline crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:519]: Having so many functions named `ext_` is making me wonder whether we want to have this in a separate `extensions` module rather than co-locating it with `kiosk`

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (sblackshear) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:446]: I would anticipate that folks will want to use an allowlist and/or a denylist in specifying which types an extension can access (just about any system I've seen for one ends up wanting the other), so we should probably include fields for both.

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:441]: Of course, we used to have a separate module in previous implementations: https://github.com/MystenLabs/sui/pull/12244/files#diff-3b0c93d3b04ef897619e3bbf58de9dacfddf69f72ff733a8c908e37a11f67b3a

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:446]: Agree that `allowed_types` fits it better. However, Extension is never available to anyone and intentionally abstracted away so that we can change the format later. Having said that, working with a vector without an Option simplifies things a lot (remember how many issues developers face with Option in a PTB).

Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:446]: >I would anticipate that folks will want to use an allowlist and/or a denylist in specifying which types an extension can access (just about any system I've seen for one ends up wanting the other), so we should probably include fields for both.

In this regard I don't think anyone would ever use denylist. The goal of this setting is to protect the Kiosk from being used on types the Kiosk Owner doesn't want to see. Eg I only want Capy Mixer extension to add a new Capy while I don't want it to add any other object. 

One can always create a new type for scammy-spammy object and having a denylist for that purpose seems like a waste of time.

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:447]: Yeah, the docs could be better.

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:458]: There's multiple reasons for having keys custom in Kiosk:

1. Easy to navigate - when I query dynamic fields I can detect extension because it starts with 0x2::kiosk::ExtensionKey, otherwise it would be a random field which may or may not mean an extension - bad devx.

2. Allowing third party keys on the base level would allow the third party read this fields avoiding the extensions functions. While it's not the case - `Extension` struct has no public getters; it's a general pattern which is coherent with other dynamic fields in the Kiosk (and there're many).

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:552]: It's a very careful implementation approach because the final set of permissions is still under discussion. For example: should `place` be a separate permission with `lock`? While `place` is innocent, `lock` may lead to an unwanted asset being stuck forever in a Kiosk. At the same time locking and placing mechanics behave symmetrically in trades.

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:497]: I didn't quite get that? If the signature is "enable" then it causes confusion around default "enabled" scenario.

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:519]: All attempts to move it outside ended nowhere, access control is very tight in Kiosk.

**Comment by damirka** [comment]
@sblackshear I've been thinking about this implementation for some time and I see a potential to make it even more generic, abstracting (generalizing) away the `Extension` from the Kiosk. Practically speaking, the only "base" feature of the extensions API is to provide "permissions" and access to protected actions.

Perhaps, we could try something of kind:
```move
module sui::kiosk {
    // ...

    struct ExtensionKey<phantom Ext> has store, copy, drop {}

    public fun add_extension<Ext: drop, Extension: store>(
        _ext: Ext,
        kiosk: &mut Kiosk,
        extension: Extension
    );

    public fun extension_mut<Ext: drop, Extension: store>(
        _ext: Ext,
        kiosk: &mut Kiosk,
    ): &mut Extension;

    public fun remove_extenstion<Ext: drop, Extension: store>(
        _ext: Ext,
        kiosk: &mut Kiosk,
    ): Extension;
}
```

Wonder what you think and also where the permissions could live. I think we could even create a dynamic field with permissions on the side. So one would be for the `Extension` and one for the `ExtensionConfig`. Principally it's the choice between having everything pre-defined + giving a Bag for custom data vs not having anything defined but permissions.

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (amnn) [inline crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:42]:
```suggestion
/// - `listed_exclusively` - An item is listed via the `list_with_purchase_cap`
```

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (amnn) [inline crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:543]: What's the rationale for weakening this restriction?

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (amnn) [inline crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:580]: This is a duplicate of `kiosk_owner_cap_for`, below.

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move
Comment (amnn) [inline crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move:97]: nit: This is useful context for the review, but I would remove reference to an "original implementation" before landing this, because it's going to be confusing for users of the API who won't know or care about the fact that there was a previous version of this code.

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move
Comment (amnn) [inline crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move:197]:
```suggestion
    /// requires a `TransferPolicy` for the placed type to exist.
```

What happens if the policy exists at the time of placing and is subsequently made inaccessible somehow?

File: kiosk/examples/unlock.move
Comment (amnn) [inline kiosk/examples/unlock.move:20]:
```suggestion
    /// Trying to destroy the Unlocker object while not being its publisher.
```

File: kiosk/examples/unlock.move
Comment (amnn) [inline kiosk/examples/unlock.move:37]: Checking my understanding: the check that `T` is from `publisher`'s package happens in `policy::new` because by type inference, we know that `policy` and `cap` must be parameterised by the same `T`?

**Review by amnn — COMMENTED** [review]
Looks good to me at a high level!  Will leave it to @sblackshear for the final review -- I don't think I have all the context to say whether it addresses customer needs.

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move
Comment (manolisliolios) [inline crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move:194]: I don't think we do any `allowed types` checks anymore!

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move:580]: Nice catch! Almost missed this one.

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move:97]: Thanks! Yeah, this one was more like a note to self or to reviewer.

File: kiosk/examples/unlock.move
Comment (damirka) [reply kiosk/examples/unlock.move:37]: Exactly! That is how it happens.

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move:66]: It provides safety when trying to remove the extension. This way we ensure there are no dangling objects (w/o a parent) + this directly aligns with protecting the assets.

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move
Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move:78]: We don't know, but having 32 is way more than there are actions today (currently, we have 6 possible actions iirc).

**Comment by damirka** [comment]
Thank you for the review! @amnn @manolisliolios @sblackshear 

I've addressed your comments, re-requesting to get a stamp if everyone agrees.

File: crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move
Comment (sblackshear) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move:78]: I'd be inclined to do 64 just to be conservative, but we can always do `Action2` if needed.

Comment (damirka) [reply crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk_extension.move:78]: My original choice was u16. 😛
Happy to make it u64.

**Review by sblackshear — APPROVED** [review]

**Review by manolisliolios — APPROVED** [review]
Awesome!

**Pull request merged**
